src/main/scala/org/scalameter/japi/JBench.scala

```diff
@@ -0,0 +1,271 @@
+package org.scalameter.japi
+
+import java.lang.annotation.Annotation
+import java.lang.reflect.Method
+import org.scalameter.BasePerformanceTest._
+import org.scalameter._
+import org.scalameter.japi.annotation._
+import org.scalameter.reporting.{HtmlReporter, RegressionReporter}
+import scala.collection.mutable
+import scala.language.reflectiveCalls
+import scala.util.{DynamicVariable, Try}
+
+
+/** Base class for all annotation based benchmarks. */
+abstract class JBench extends BasePerformanceTest with Serializable {
+
+  override final def rebuildSetupZipper() = constructSetupTree()
+
+  /** Constructs setup tree.
+   */
+  private def constructSetupTree() = {
+    object BenchmarkExtractor {
+      def unapply(m: Method): Option[Seq[String]] = {
+        Option(m.getAnnotation(classOf[benchmark])).map(_.value().split('.').toSeq)
+      }
+    }
+
+    val clazz = this.getClass
+    val scopedSetups: Seq[(Seq[String], Seq[Method])] = {
+      val mapping = mutable.Map.empty[Seq[String], Seq[Method]]
+      for (m @ BenchmarkExtractor(scopes) <- clazz.getMethods) {
+        mapping.get(scopes) match {
+          case Some(setups) =>
+            mapping += (scopes -> (setups :+ m))
+          case None =>
+            mapping += (scopes -> Seq(m))
+        }
+      }
+      mapping.toList
+    }
+    val scopeCtxs = {
+      Option(clazz.getAnnotation(classOf[scopes])).map { a =>
+        a.value().map { sc =>
+          sc.scope().split('.').toSeq -> getField[Context](clazz, sc.context())
+        }.toMap
+      }.getOrElse(Map.empty)
+    }
+    for ((scope, mapping) <- scopedSetups.groupBy(_ ._1.head)) {
+      setScope(clazz, scopeCtxs, scope, mapping.map(kv => (kv._1.tail, kv._2)))
+    }
+  }
+
+  /** Sets given scope.
```

Comment:
Extends the scope named `name` with the corresponding context specified in `scopeCtxs` and adds the corresponding benchmark snippets in `scopedSetups`. After this method is invoked, the setup zipper is set to a setup tree that contains all the benchmark snippets.